### PR TITLE
stdlib: build multiple platform libraries simultaneously

### PR DESCRIPTION
**Pull request by compnerd**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

Always construct the platform libraries for all the target SDKs.  This enables
us to build the right set for targets and will also enable us to cross-compile
multiple targets simultaneously.

### CODE REVIEW

**Comment by compnerd** [comment]
CC @gribozavr

**Comment by compnerd** [comment]
CC @modocache

**Comment by modocache** [comment]
Fantastic, thanks! I'll wait to hear what @gribozavr thinks before merging myself.

@swift-ci please test

File: stdlib/public/Platform/CMakeLists.txt [hunk 1]
```diff
@@ -6,28 +6,21 @@ set(swift_platform_sources
...
-      DEPENDS glibc_modulemap)
-endif()
+add_swift_library(swiftDarwin ${SWIFT_STDLIB_LIBRARY_BUILD_TYPES} IS_SDK_OVERLAY
+    Darwin.swift
+    ${swift_platform_sources}
+    POSIXError.swift
+    MachError.swift
+
+    SWIFT_COMPILE_FLAGS -Xfrontend -disable-objc-attr-requires-foundation-module
+    TARGET_SDKS IOS;OSX;TVOS;WATCHOS
```
Comment (gribozavr) [inline stdlib/public/Platform/CMakeLists.txt]: The syntax of `TARGET_SDKS` does not require semicolons, see, for example, `stdlib/public/SDK/CloudKit/CMakeLists.txt`.

File: stdlib/private/SwiftPrivateLibcExtras/CMakeLists.txt [hunk 2]
```diff
@@ -1,21 +1,16 @@
...
-if(SWIFT_HOST_VARIANT MATCHES "${SWIFT_DARWIN_VARIANTS}")
-  list(APPEND swift_private_darwin_extras_module_depends
-    Darwin)
-elseif(SWIFT_HOST_VARIANT STREQUAL "linux" OR SWIFT_HOST_VARIANT STREQUAL "freebsd")
-  list(APPEND swift_private_darwin_extras_module_depends
-    Glibc)
-endif()
-
-add_swift_library(swiftSwiftPrivateLibcExtras ${SWIFT_STDLIB_LIBRARY_BUILD_TYPES} IS_STDLIB
+add_swift_library(swiftSwiftPrivateLibcExtras ${SWIFT_STDLIB_LIBRARY_BUILD_TYPES} TARGET_LIBRARY IS_STDLIB
```
Comment (gribozavr) [inline stdlib/private/SwiftPrivateLibcExtras/CMakeLists.txt]: I thought `IS_STDLIB` implies `TARGET_LIBRARY`, doesn't that work?

```cmake
# From cmake/modules/AddSwift.cmake

  # Standard library is always a target library.
  if(SWIFTLIB_IS_STDLIB)
    set(SWIFTLIB_HAS_SWIFT_CONTENT TRUE)
    set(SWIFTLIB_TARGET_LIBRARY TRUE)
  endif()
```

**Comment by gribozavr** [comment]
CI error:

```
ninja: error: '/home/buildnode/jenkins/workspace/swift-PR-Linux/swift/stdlib/private/SwiftPrivateLibcExtras/swiftSWIFT_MODULE_DEPENDS_FreeBSD-linux-x86_64', needed by 'stdlib/private/SwiftPrivateLibcExtras/linux/x86_64/SwiftPrivateLibcExtras.o', missing and no known rule to make it
```

File: stdlib/private/SwiftPrivateLibcExtras/CMakeLists.txt [hunk 3]
```diff
@@ -1,21 +1,16 @@
...
 
-  SWIFT_MODULE_DEPENDS ${swift_private_darwin_extras_module_depends}
+  SWIFT_MODULE_DEPENDS SwiftPrivate
   SWIFT_COMPILE_FLAGS ${STDLIB_SIL_SERIALIZE_ALL}
+  SWIFT_MODULE_DEPENDS_OSX Darwin
+  SWIFT_MODULE_DEPENDS_IOS Darwin
+  SWIFT_MODULE_DEPENDS_TVOS Darwin
+  SWIFT_MODULE_DEPENDS_WATCHOS Darwin
+  SWIFT_MODULE_DEPENDS_LINUX Glibc
+  SWIFT_MODULE_DEPENDS_FreeBSD Glibc
```
Comment (gribozavr) [inline stdlib/private/SwiftPrivateLibcExtras/CMakeLists.txt]: `FREEBSD` should be all caps.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@gribozavr addressed the comments, added a few more libraries!

**Comment by gribozavr** [comment]
@compnerd CI failure:

```
ninja: error: '/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/stdlib/public/SDK/simd/swiftDarwin-iphonesimulator-i386', needed by 'stdlib/public/SDK/simd/iphonesimulator/i386/simd.o', missing and no known rule to make it
```

**Comment by compnerd** [comment]
@gribozavr oh wow, thats a slightly annoying difference.  SWIFT_MODULE_DEPENDS IOS covers both IOS and IOS_SIMULATOR, however, TARGET_SDKS does not.  Oh well, explicitly listed.  Hopefully this actually goes through this time.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@gribozavr the Linux failure is unrelated (was fixed by Chris recently).  The OS X build failure is confusing me though.  Im not sure why this change would suddenly prevent module imports from functioning.

**Comment by modocache** [comment]
I'm planning on taking a look at the OS X failure myself, but let's confirm that Linux works now:

@swift-ci please test

File: stdlib/private/SwiftPrivatePthreadExtras/CMakeLists.txt [hunk 4]
```diff
@@ -1,20 +1,14 @@
...
   # filename.
   SwiftPrivatePthreadExtras.swift
   PthreadBarriers.swift
 
-  SWIFT_MODULE_DEPENDS ${swift_private_pthread_extras_module_depends}
+  SWIFT_MODULE_DEPENDS_IOS Darwin
+  SWIFT_MODULE_DEPENDS_OSX Darwin
+  SWIFT_MODULE_DEPENDS_WATCHOS Darwin
+  SWIFT_MODULE_DEPENDS_LINUX Glibc
+  SWIFT_MODULE_DEPENDS_FREEBSD Glibc
```
Comment (rintaro) [inline stdlib/private/SwiftPrivatePthreadExtras/CMakeLists.txt]: `TVOS`?

Comment (compnerd) [inline stdlib/private/SwiftPrivatePthreadExtras/CMakeLists.txt]: *sigh* That is shameful.  Thanks @rintaro, thats an amazing eye you have there!

**Comment by compnerd** [comment]
@swift-ci Please test

**Comment by compnerd** [comment]
@swift-ci Please test OS X platform

**Comment by compnerd** [comment]
Thanks to @rintaro, I think that this is good now!

**Comment by compnerd** [comment]
@gribozavr okay to merge?

File: stdlib/private/StdlibUnittest/CMakeLists.txt [hunk 5]
```diff
@@ -1,31 +1,19 @@
...
-endif()
 if(SWIFT_SERIALIZE_STDLIB_UNITTEST)
   list(APPEND swift_stdlib_unittest_compile_flags "-Xfrontend" "-sil-serialize-all")
 endif()
 
+# TODO: extend this through a similar mechanism to SWIFT_MODULE_DEPENDS
+set(swift_stdlib_unittest_framework_depends)
+if(SWIFT_HOST_VARIANT MATCHES "${SWIFT_DARWIN_VARIANTS}")
+  list(APPEND swift_stdlib_unittest_framework_depends Foundation)
+endif()
```
Comment (gribozavr) [inline stdlib/private/StdlibUnittest/CMakeLists.txt]: I'm not sure if this is actually needed, might be an old workaround for something.  StdlibUnittest imports the Foundation overlay, so there should be no need to list Foundation as an explicit dependency.  You could try removing this block in a future PR.

**Comment by gribozavr** [comment]
LGTM, let's merge it when the CI is green.

**Comment by compnerd** [comment]
@swift-ci Please test and merge

**Comment by compnerd** [comment]
Seems that Jenkins is having some locking issues:

https://ci.swift.org/job/swift-PR-Linux/2558/
https://ci.swift.org/job/swift-PR-osx/2260/ 

Its stuck at "Github Pull Request Merger is waiting for a checkpoint on Swift Test and Merge #630".

Just going to manually merge it since CI went green.

**Pull request merged**